pkgs/applications/networking/browsers/mullvad-browser/default.nix

```diff
@@ -146,7 +170,12 @@ stdenv.mkDerivation rec {
     categories = [ "Network" "WebBrowser" "Security" ];
   })];
 
-  buildPhase = ''
+  buildPhase = if stdenv.isDarwin then ''
```

Comment:
The `buildPhase` does quite a lot of things for Linux and is basically skipped completely here for macOS.

`touch "$MB_IN_STORE/system-install"`
Tor Browser, at least on Linux, is a portable package and stores its state within its own folder.
Creating this `system-install` file causes it to store its state in a user directory (`~/.mullvad/mullvadbrowser` on Linux) instead. Not sure how this works on macOS, perhaps you could tell me where it stores its state with the current setup.

`FONTCONFIG_FILE=$MB_IN_STORE/fontconfig/fonts.conf`
We need it in the nix package, because we don't invoke the `start-mullvad-browser` wrapper directly.
`start-mullvad-browser` sets up the environment & adds some functionality like `--register-app`/`--unregister-app` and then launches the actual binary. This doesn't fully work on NixOS, so we have our own wrapping where we need to make sure to set important variables like `FONTCONFIG_FILE` to avoid leaking fonts & thus changing the browesr fingerprint.

Then there are a couple of settings, some of those are to avoid capturing nix store paths, some change some defaults like `noscript.firstRunRedirection`.

I hope this already gives you a bit of context about this.
Generally, I think we should try to make the Linux & macOS configurations the same except for OS-specific configuration (pulseaudio, etc.).

Reply:
> touch "$MB_IN_STORE/system-install"
> Tor Browser, at least on Linux, is a portable package and stores its state within its own folder.
Creating this system-install file causes it to store its state in a user directory (~/.mullvad/mullvadbrowser on Linux) instead. Not sure how this works on macOS, perhaps you could tell me where it stores its state with the current setup.

The comparable `~/.mullvad/mullvadbrowser` directory is `~/Library/Application Support/MullvadBrowser` on MacOS:

<img width="447" alt="image" src="https://github.com/NixOS/nixpkgs/assets/18178614/04e8a667-ff2c-4bf7-8082-3b637c4dbf8d">

Is this `system-install` behavior documented anywhere? My first assumption is that it doesn't operate the same on MacOS.

---

The other changes you point out in `buildPhase` -- in my mind -- wouldn't translate to MacOS (i.e. since MacOS apps are self-contained, MacOS doesn't handle fonts the same way Linux does, etc.), although, I appreciate the context and its helpful nonetheless. Would you like me to more fully investigate these and try to copy the same changes in the build?

Reply:
> Is this system-install behavior documented anywhere? My first assumption is that it doesn't operate the same on MacOS.

Not really, I had quite a few issues even finding & understanding that functionality.
I originally found it used by the Flatpak: https://github.com/flathub/net.mullvad.MullvadBrowser/blob/27161259ff26a28efc839c726ebde9a4d4b4ac29/net.mullvad.MullvadBrowser.yml#L67

However you can find some mentions of `system-install` in Tor Project's GitLab including the source code using it:
https://gitlab.torproject.org/search?scope=issues&search=system-install
https://gitlab.torproject.org/search?search=%22system-install%22&nav_source=navbar&project_id=472&group_id=477&search_code=true&repository_ref=tor-browser-115.7.0esr-13.5-1

And these config files for Mullvad Browser & Tor Browser could also be of interest, as they define the directories:
https://gitlab.torproject.org/tpo/applications/mullvad-browser/-/blob/mullvad-browser-115.6.0esr-13.5-1/browser/config/mozconfigs/mullvad-browser?ref_type=heads
https://gitlab.torproject.org/tpo/applications/tor-browser/-/blob/tor-browser-115.7.0esr-13.5-1/browser/config/mozconfigs/tor-browser?ref_type=heads

> The comparable `~/.mullvad/mullvadbrowser` directory is `~/Library/Application Support/MullvadBrowser` on MacOS

Does it already use that directory without the `system-install` file? In that case the `system-install` is probably not necessary.

Reply:
> Does it already use that directory without the system-install file? 

I believe so, but is there any way I can verify this conclusively? I would imagine that the Tor Browser/Mullvad Browser is not as ephemeral/self-contained on MacOS as it is on Linux and as a result uses that directory by default.

Reply:
> The other changes you point out in `buildPhase` -- in my mind -- wouldn't translate to MacOS (i.e. since MacOS apps are self-contained. [...] Would you like me to more fully investigate these and try to copy the same changes in the build?

If possible, it would be nice if `extraPrefs` also works on macOS. And then we could also set the `extensions.xpiState` & `noscript.firstRunRedirection` config values.

Reply:
> I believe so, but is there any way I can verify this conclusively?

If you can see the `profiles.ini` & profile directory in there I think that's proof enough.
I don't remember properly, but on Linux without `system-install` file it would try to store its state in the directory of the binary or the working directory or something like that.
As long as it always stores its state in the same place, I don't see an issue here.